Testcafe tests failing frequently, but inconsistently

Master build failing; seems to have gotten flakier after #367

The test that causes the failure also appears to be random; here are some of the ones that are failing:

https://travis-ci.com/github/internetarchive/bookreader/builds/182263172 
```
 Base Tests for right to left book: gendaitankashu00meijuoft
...
 ✖ Clicking `page flip buttons` updates location
   1) AssertionError: expected false to be truthy
      Browser: Firefox 80.0 / Linux 0.0
         206 |  test('Clicking `page flip buttons` updates location', async
      t => {
         207 |    const { nav } = br;
         208 |    const pageProgression = await bookPageProgression();
         209 |    // Page navigation creates params
         210 |    await t.click(getNext(nav, pageProgression))
       > 211 |      .expect(expectPage()).ok()
         212 |      .expect(expectMode('2up')).ok();
         213 |
         214 |    await t.click(getPrevious(nav, pageProgression))
         215 |      .expect(expectPage()).ok()
         216 |      .expect(expectMode('2up')).ok();
         at <anonymous>
     (/home/travis/build/internetarchive/bookreader/tests/e2e/helpers/base.js:211:29)
```

https://travis-ci.com/github/internetarchive/bookreader/builds/182263044

```
 Base Tests for: theworksofplato01platiala
...
 ✖ 2up mode - Clicking `Previous page` changes the page
    1) AssertionError: expected
      'https://ia600900.us.archive.org/BookReader/BookReaderImages.php?zip=/25/items/theworksofplato01platiala/theworksofplato01platiala_jp2.zip&file=theworksofplato01platiala_jp2/theworksofplato01platiala_0009.jp2&id=theworksofplato01platiala&scale=4&rotate=0'
         to not deeply equal
         'https://ia600900.us.archive.org/BookReader/BookReaderImages.php?zip=/25/items/theworksofplato01platiala/theworksofplato01platiala_jp2.zip&file=theworksofplato01platiala_jp2/theworksofplato01platiala_0009.jp2&id=theworksofplato01platiala&scale=4&rotate=0'
         Browser: Firefox 80.0 / Linux 0.0
               163 |
                  164 |    // we aren't showing the same image in both leaves
                  165 |    await t.expect(origImg1Src).notEql(origImg2Src);
                  166 |
                  167 |    // we are showing new pages
                > 168 |    await t.expect(prevImg1Src).notEql(origImg1Src);
                169 |    await t.expect(prevImg1Src).notEql(origImg2Src);
                  170 |    await t.expect(prevImg2Src).notEql(origImg1Src);
                  171 |    await t.expect(prevImg2Src).notEql(origImg2Src);
                  172 |
                  173 |    // we aren't showing the same image in the new pages
                  at <anonymous>
               (/home/travis/build/internetarchive/bookreader/tests/e2e/helpers/base.js:168:33)
```

```
 Base Tests for right to left book: gendaitankashu00meijuoft
...
 ✖ Clicking `page flip buttons` updates location
    1) An error occurred in ClientFunction code:
         
            TypeError: window.br is undefined
            Browser: Firefox 80.0 / Linux 0.0
               203 |    await t.expect(nextImg1Src).notEql(nextImg2Src);
                  204 |  })
                  205 |
                  206 |  test('Clicking `page flip buttons` updates location', async
               t => {
                   207 |    const { nav } = br;
                > 208 |    const pageProgression = await bookPageProgression();
                209 |    // Page navigation creates params
                  210 |    await t.click(getNext(nav, pageProgression))
                  211 |      .expect(expectPage()).ok()
                  212 |      .expect(expectMode('2up')).ok();
                  213 |
                  at <anonymous>
            (/home/travis/build/internetarchive/bookreader/tests/e2e/helpers/base.js:208:35)
            at <anonymous>
            (/home/travis/build/internetarchive/bookreader/tests/e2e/helpers/base.js:206:3)
            at <anonymous>
            (/home/travis/build/internetarchive/bookreader/node_modules/testcafe/src/api/wrap-test-function.js:17:28)
            at TestRun._executeTestFn
            (/home/travis/build/internetarchive/bookreader/node_modules/testcafe/src/test-run/index.js:296:19)
            at TestRun.start
               (/home/travis/build/internetarchive/bookreader/node_modules/testcafe/src/test-run/index.js:346:24)
```



https://travis-ci.com/github/internetarchive/bookreader/builds/182458827
```
 Base Tests for: theworksofplato01platiala
...
 ✖ 2up mode - Clicking `Previous page` changes the page
    1) AssertionError: expected
      'https://ia800900.us.archive.org/BookReader/BookReaderImages.php?zip=/25/items/theworksofplato01platiala/theworksofplato01platiala_jp2.zip&file=theworksofplato01platiala_jp2/theworksofplato01platiala_0006.jp2&id=theworksofplato01platiala&scale=4&rotate=0'
         to not deeply equal
         'https://ia800900.us.archive.org/BookReader/BookReaderImages.php?zip=/25/items/theworksofplato01platiala/theworksofplato01platiala_jp2.zip&file=theworksofplato01platiala_jp2/theworksofplato01platiala_0006.jp2&id=theworksofplato01platiala&scale=4&rotate=0'
         
            + expected - actual
            Browser: Firefox 80.0 / Linux 0.0
               163 |
                  164 |    // we aren't showing the same image in both leaves
                  165 |    await t.expect(origImg1Src).notEql(origImg2Src);
                  166 |
                  167 |    // we are showing new pages
                > 168 |    await t.expect(prevImg1Src).notEql(origImg1Src);
                169 |    await t.expect(prevImg1Src).notEql(origImg2Src);
                  170 |    await t.expect(prevImg2Src).notEql(origImg1Src);
                  171 |    await t.expect(prevImg2Src).notEql(origImg2Src);
                  172 |
                  173 |    // we aren't showing the same image in the new pages
                  at <anonymous>
               (/home/travis/build/internetarchive/bookreader/tests/e2e/helpers/base.js:168:33)
```

--- a/.testcaferc.json
+++ b/.testcaferc.json
@@ -1,5 +1,5 @@
 {
   "appCommand": "npm run serve",
-  "browsers": ["chrome", "firefox"],
+  "browsers": ["chrome:headless", "firefox:headless"],
   "src": ["tests/e2e/**/*test.js"]
 }

--- a/.travis.yml
+++ b/.travis.yml
@@ -15,5 +15,6 @@ script:
   - npm run lint
   - npm run test:coverage
   # Travis machines usually have 2 cores (see https://docs.travis-ci.com/user/reference/overview/#virtualisation-environment-vs-operating-system )
-  - BASE_URL=http://127.0.0.1:8000/BookReaderDemo/ npx testcafe "chrome:headless,firefox:headless" --concurrency 2
+  # Quarantine mode (sigh) to catch unstable tests (see https://devexpress.github.io/testcafe/documentation/guides/basic-guides/run-tests.html#quarantine-mode )
+  - BASE_URL=http://127.0.0.1:8000/BookReaderDemo/ npx testcafe "chrome:headless,firefox:headless" --concurrency 2 --quarantine-mode
   - npm run codecov

--- a/.travis.yml
+++ b/.travis.yml
@@ -16,5 +16,5 @@ script:
   - npm run test:coverage
   # Travis machines usually have 2 cores (see https://docs.travis-ci.com/user/reference/overview/#virtualisation-environment-vs-operating-system )
   # Quarantine mode (sigh) to catch unstable tests (see https://devexpress.github.io/testcafe/documentation/guides/basic-guides/run-tests.html#quarantine-mode )
-  - BASE_URL=http://127.0.0.1:8000/BookReaderDemo/ npx testcafe "chrome:headless,firefox:headless" --concurrency 2 --quarantine-mode
+  - BASE_URL=http://127.0.0.1:8000/BookReaderDemo/ npx testcafe --concurrency 2 --quarantine-mode
   - npm run codecov

--- a/BookReaderDemo/demo-internetarchive.html
+++ b/BookReaderDemo/demo-internetarchive.html
@@ -55,10 +55,6 @@
 // Override options coming from IA
 BookReader.optionOverrides.imagesBaseURL = '/BookReader/images/';
 BookReader.optionOverrides.enableTextSelection = true;
-
-// Temporarily fix build by mocking SearchInsideAccess; long term fix likely needs to be
-// on petabox side
-function SearchInsideAccess(br, identifier, callback) { this.init = callback; }
 </script>
 
 <script id="pageUrl" type="text/javascript"></script>

--- a/src/js/BookReader/Navbar/Navbar.js
+++ b/src/js/BookReader/Navbar/Navbar.js
@@ -116,6 +116,7 @@ export class Navbar {
   init() {
     const { br } = this;
     const { navbarTitle: title } = br.options;
+    const isRTL = br.pageProgression === 'rl';
 
     br.refs.$BRfooter = this.$root = $(`<div class="BRfooter"></div>`);
     br.refs.$BRnav = this.$nav = $(
@@ -132,13 +133,13 @@ export class Navbar {
                 <p><span class='BRcurrentpage'></span></p>
               </li>
               <li>
-                <button class="BRicon book_left" title="Flip left">
+                <button class="BRicon book_left ${isRTL ? 'book_flip_next' : 'book_flip_prev'}" title="Flip left">
                   <div class="icon icon-left-arrow"></div>
                   <span class="tooltip">Flip left</span>
                 </button>
               </li>
               <li>
-                <button class="BRicon book_right" title="Flip right">
+                <button class="BRicon book_right ${isRTL ? 'book_flip_prev' : 'book_flip_next'}" title="Flip right">
                   <div class="icon icon-left-arrow hflip"></div>
                   <span class="tooltip">Flip right</span>
                 </button>

--- a/tests/e2e/helpers/base.js
+++ b/tests/e2e/helpers/base.js
@@ -8,7 +8,7 @@ const getUrl = ClientFunction(() => window.location.href);
 /**
  * Check URL page paramter in # and path
  */
-const expectPage = ClientFunction(() => {
+const isPageInUrl = ClientFunction(() => {
   const hash = document.location.hash;
   if (hash) {
     return hash.indexOf('#page/') > -1;
@@ -22,7 +22,7 @@ const expectPage = ClientFunction(() => {
  *
  * @param mode '1up', '2up', 'thumb'
  */
-const expectMode = ClientFunction((mode) => {
+const isModeInUrl = ClientFunction((mode) => {
   const hash = document.location.hash;
   if (hash) {
     return hash.indexOf('/mode/' + mode) > -1;
@@ -87,16 +87,16 @@ export function runBaseTests (br) {
     const initialUrl = await getUrl();
 
     // Set Cookie by page navigation, wait for cookie
-    await t.click(nav.desktop.goNext)
-      .wait(PAGE_FLIP_WAIT_TIME);
+    await t.click(nav.desktop.goNext);
+    await t.wait(PAGE_FLIP_WAIT_TIME);
 
     // reload canonical URL, wait for URL change
-    await t.navigateTo(initialUrl)
-      .wait(PAGE_FLIP_WAIT_TIME);
+    await t.navigateTo(initialUrl);
+    await t.wait(PAGE_FLIP_WAIT_TIME);
 
     if (await usesResume()) {
-      await t.expect(expectPage()).ok(initialUrl)
-        .expect(expectMode('2up')).ok();
+      await t.expect(isPageInUrl()).eql(true, initialUrl);
+      await t.expect(isModeInUrl('2up')).eql(true, initialUrl)
     } else {
       // No plugin, no br-resume cookie
       await t.expect(getUrl()).notContains('#page/');
@@ -110,7 +110,7 @@ export function runBaseTests (br) {
 
     // Go to next page, so we can go previous if at front cover
     await t.click(nav.desktop.goNext);
-    await t.wait(PAGE_FLIP_WAIT_TIME); // wait for animation and page flip to happen
+    await t.wait(PAGE_FLIP_WAIT_TIME);
 
     const onLoadBrState = BRcontainer.child(0);
     const initialImages = onLoadBrState.find('img');
@@ -118,7 +118,7 @@ export function runBaseTests (br) {
     const origImg2Src = await initialImages.nth(-1).getAttribute('src');
 
     await t.click(nav.desktop.goPrev);
-    await t.wait(PAGE_FLIP_WAIT_TIME); // wait for animation and page flip to happen
+    await t.wait(PAGE_FLIP_WAIT_TIME);
 
     const nextBrState = await Selector('.BRcontainer').child(0);
     const prevImages = nextBrState.find('img');
@@ -147,7 +147,7 @@ export function runBaseTests (br) {
     const origImg2Src = await initialImages.nth(-1).getAttribute('src');
 
     await t.click(nav.desktop.goNext);
-    await t.wait(PAGE_FLIP_WAIT_TIME); // wait for animation and page flip to happen
+    await t.wait(PAGE_FLIP_WAIT_TIME);
 
     const nextBrState = await Selector('.BRcontainer').child(0);
     const nextImages = nextBrState.find('img');
@@ -170,13 +170,13 @@ export function runBaseTests (br) {
   test('Clicking `page flip buttons` updates location', async t => {
     const { nav } = br;
     // Page navigation creates params
-    await t.click(nav.desktop.goNext)
-      .expect(expectPage()).ok()
-      .expect(expectMode('2up')).ok();
+    await t.click(nav.desktop.goNext);
+    await t.expect(isPageInUrl()).eql(true);
+    await t.expect(isModeInUrl('2up')).eql(true);
 
-    await t.click(nav.desktop.goPrev)
-      .expect(expectPage()).ok()
-      .expect(expectMode('2up')).ok();
+    await t.click(nav.desktop.goPrev);
+    await t.expect(isPageInUrl()).eql(true);
+    await t.expect(isModeInUrl('2up')).eql(true);
   });
 
   test('Clicking `2 page view` brings up 2 pages at a time', async t => {

--- a/tests/e2e/helpers/base.js
+++ b/tests/e2e/helpers/base.js
@@ -32,40 +32,9 @@ const expectMode = ClientFunction((mode) => {
 });
 
 /**
- * Gets `pageProgression` value from BookReader instance
- */
-const bookPageProgression = ClientFunction(() => {
-  return window.br.pageProgression;
-})
-
-/**
- * Gets next button, depending on book's page progression
- *
- * @param [DesktopNav] navModel
- * @param [string] pageProgression - rl | lr
- * @return [Selector] "next button"
- */
-const getNext = (navModel, pageProgression = 'lr') => {
-  const r2l = pageProgression === 'rl';
-  return r2l ? navModel.desktop.goLeft : navModel.desktop.goRight;
-};
-
-/**
- * Gets previous button, depending on book's page progression
- *
- * @param [Navigation] navModel
- * @param [string] pageProgression - rl | lr
- * @return [Selector] "previous button"
- */
-const getPrevious = (navModel, pageProgression = 'lr') => {
-  const r2l = pageProgression === 'rl';
-  return r2l ? navModel.desktop.goRight : navModel.desktop.goLeft;
-};
-
-/**
  * Runs all expected base tests for BookReader
  *
- * @param { BookReader } br - Model
+ * @param { import('../models/BookReader').default } br - Model
  */
 export function runBaseTests (br) {
   test('On load, pages fit fully inside of the BookReader™', async t => {
@@ -118,9 +87,7 @@ export function runBaseTests (br) {
     const initialUrl = await getUrl();
 
     // Set Cookie by page navigation, wait for cookie
-    const pageProgression = await bookPageProgression();
-    const goNext = getNext(nav, pageProgression);
-    await t.click(goNext)
+    await t.click(nav.desktop.goNext)
       .wait(PAGE_FLIP_WAIT_TIME);
 
     // reload canonical URL, wait for URL change
@@ -141,11 +108,8 @@ export function runBaseTests (br) {
   test('2up mode - Clicking `Previous page` changes the page', async t => {
     const { nav, BRcontainer} = br;
 
-    const pageProgression = await bookPageProgression();
-    const goPrevious = getPrevious(nav, pageProgression);
-    const goNext = getNext(nav, pageProgression);
     // Go to next page, so we can go previous if at front cover
-    await t.click(goNext);
+    await t.click(nav.desktop.goNext);
     await t.wait(PAGE_FLIP_WAIT_TIME); // wait for animation and page flip to happen
 
     const onLoadBrState = BRcontainer.child(0);
@@ -153,7 +117,7 @@ export function runBaseTests (br) {
     const origImg1Src = await initialImages.nth(0).getAttribute('src');
     const origImg2Src = await initialImages.nth(-1).getAttribute('src');
 
-    await t.click(goPrevious);
+    await t.click(nav.desktop.goPrev);
     await t.wait(PAGE_FLIP_WAIT_TIME); // wait for animation and page flip to happen
 
     const nextBrState = await Selector('.BRcontainer').child(0);
@@ -182,7 +146,7 @@ export function runBaseTests (br) {
     const origImg1Src = await initialImages.nth(0).getAttribute('src');
     const origImg2Src = await initialImages.nth(-1).getAttribute('src');
 
-    await t.click(getNext(nav, await bookPageProgression()));
+    await t.click(nav.desktop.goNext);
     await t.wait(PAGE_FLIP_WAIT_TIME); // wait for animation and page flip to happen
 
     const nextBrState = await Selector('.BRcontainer').child(0);
@@ -205,13 +169,12 @@ export function runBaseTests (br) {
 
   test('Clicking `page flip buttons` updates location', async t => {
     const { nav } = br;
-    const pageProgression = await bookPageProgression();
     // Page navigation creates params
-    await t.click(getNext(nav, pageProgression))
+    await t.click(nav.desktop.goNext)
       .expect(expectPage()).ok()
       .expect(expectMode('2up')).ok();
 
-    await t.click(getPrevious(nav, pageProgression))
+    await t.click(nav.desktop.goPrev)
       .expect(expectPage()).ok()
       .expect(expectMode('2up')).ok();
   });

--- a/tests/e2e/models/Navigation.js
+++ b/tests/e2e/models/Navigation.js
@@ -18,17 +18,28 @@ export default class Navigation {
  */
 class DesktopNav {
   constructor(bottomToolbar, topToolbar) {
+    // flipping
     this.goLeft = bottomToolbar.find('.BRicon.book_left');
     this.goRight = bottomToolbar.find('.BRicon.book_right');
+    this.goNext = bottomToolbar.find('.BRicon.book_flip_next');
+    this.goPrev = bottomToolbar.find('.BRicon.book_flip_prev');
+
+    // mode switching
     this.mode1Up = bottomToolbar.find('.BRicon.onepg');
     this.mode2Up = bottomToolbar.find('.BRicon.twopg');
     this.modeThumb = bottomToolbar.find('.BRicon.thumb');
     this.viewmode = bottomToolbar.find('.BRicon.viewmode');
+
+    // zoom
     this.zoomIn = bottomToolbar.find('.BRicon.zoom_in');
     this.zoomOut = bottomToolbar.find('.BRicon.zoom_out');
-    this.fullScreen = bottomToolbar.find('.BRicon.full');
+
+    // search
     this.searchBox = topToolbar.find('.BRbooksearch.desktop');
     this.searchPin = bottomToolbar.find('.BRsearch');
+
+    // other
+    this.fullScreen = bottomToolbar.find('.BRicon.full');
     this.sliderRange = bottomToolbar.find('.ui-slider-range');
   }
 }

